### PR TITLE
OVMF fix building without network support

### PR DESCRIPTION
**Pull request by Joursoir**
# Description

It's currently not possible to compile OVMF x64 without the network stack. My patch fixes that.

The second commit fixes the build issue specifically for X64 QEMU. However, similar problems occur on other QEMU platforms (IA32, MicrovmX64, etc.) and ArmVirtPkg. I am interested in fixing these platforms as well. Moving PCDs to conditional expressions across all platforms doesn't seem like an ideal solution. Instead, I propose moving them to `NetworkPcds.dsc.inc`. Any thoughts?

- [ ] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - Examples: Does it add a new library class or move a module to a different repo.
- [ ] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [ ] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - Examples: Unit tests or integration tests.

## How This Was Tested

Tested on QEMU 9.0.2 with NETWORK_ENABLE=FALSE and NETWORK_ENABLE=TRUE builds.

## Integration Instructions

N/A

### CODE REVIEW

File: NetworkPkg/NetworkPcds.dsc.inc [hunk 1]
```diff
@@ -11,6 +11,8 @@
 #
 ##
 
+!if $(NETWORK_ENABLE) == TRUE
```
Comment (SaloniKasbekar) [inline NetworkPkg/NetworkPcds.dsc.inc:14]: Not sure if this is the right place to put it. Ideally you would check if network was enabled before including the NetworkPkg.dsc which includes this file, right?

Comment (Joursoir) [reply NetworkPkg/NetworkPcds.dsc.inc:14]: Yeah, your suggestion is the ideal case. However, note that many of the Network includes are currently being added without any checks. In this context, then we also need to make sure that the network allows HTTP secure connections (`NETWORK_ALLOW_HTTP_CONNECTIONS`) NOT within this file.

If we take the approach of checking before includes, we would have to modify more than five DSC files:

```
[triste@reeva edk2]$ grep -rni -A 2 -B 2 "\!include NetworkPkg/NetworkPcds.dsc.inc" *
grep: warning: stray \ before !
ArmVirtPkg/ArmVirtQemuKernel.dsc-160-  # Network Pcds
ArmVirtPkg/ArmVirtQemuKernel.dsc-161-  #
ArmVirtPkg/ArmVirtQemuKernel.dsc:162:!include NetworkPkg/NetworkPcds.dsc.inc
ArmVirtPkg/ArmVirtQemuKernel.dsc-163-
ArmVirtPkg/ArmVirtQemuKernel.dsc-164-  gEfiMdeModulePkgTokenSpaceGuid.PcdResetOnMemoryTypeInformationChange|FALSE
--
ArmVirtPkg/ArmVirtQemu.dsc-195-  # Network Pcds
ArmVirtPkg/ArmVirtQemu.dsc-196-  #
ArmVirtPkg/ArmVirtQemu.dsc:197:!include NetworkPkg/NetworkPcds.dsc.inc
ArmVirtPkg/ArmVirtQemu.dsc-198-
ArmVirtPkg/ArmVirtQemu.dsc-199-  # System Memory Base -- fixed at 0x4000_0000
--
NetworkPkg/NetworkPcds.dsc.inc-3-#
NetworkPkg/NetworkPcds.dsc.inc-4-# This file can be included to the [Pcds*] section(s) of a platform DSC file
NetworkPkg/NetworkPcds.dsc.inc:5:# by using "!include NetworkPkg/NetworkPcds.dsc.inc" to specify PCD settings
NetworkPkg/NetworkPcds.dsc.inc-6-# according to the value of flags described in "NetworkDefines.dsc.inc".
NetworkPkg/NetworkPcds.dsc.inc-7-#
--
NetworkPkg/Network.dsc.inc-27-
NetworkPkg/Network.dsc.inc-28-[PcdsFixedAtBuild]
NetworkPkg/Network.dsc.inc:29:!include NetworkPkg/NetworkPcds.dsc.inc
NetworkPkg/Network.dsc.inc-30-
NetworkPkg/Network.dsc.inc-31-[LibraryClasses]
--
OvmfPkg/Bhyve/BhyveX64.dsc-507-  # Network Pcds
OvmfPkg/Bhyve/BhyveX64.dsc-508-  #
OvmfPkg/Bhyve/BhyveX64.dsc:509:!include NetworkPkg/NetworkPcds.dsc.inc
OvmfPkg/Bhyve/BhyveX64.dsc-510-
OvmfPkg/Bhyve/BhyveX64.dsc-511-  # Point to the MdeModulePkg/Application/UiApp/UiApp.inf
--
OvmfPkg/OvmfPkgIa32X64.dsc-590-  # Network Pcds
OvmfPkg/OvmfPkgIa32X64.dsc-591-  #
OvmfPkg/OvmfPkgIa32X64.dsc:592:!include NetworkPkg/NetworkPcds.dsc.inc
OvmfPkg/OvmfPkgIa32X64.dsc-593-
OvmfPkg/OvmfPkgIa32X64.dsc-594-  gEfiShellPkgTokenSpaceGuid.PcdShellFileOperationSize|0x20000
--
OvmfPkg/CloudHv/CloudHvX64.dsc-558-  # Network Pcds
OvmfPkg/CloudHv/CloudHvX64.dsc-559-  #
OvmfPkg/CloudHv/CloudHvX64.dsc:560:!include NetworkPkg/NetworkPcds.dsc.inc
OvmfPkg/CloudHv/CloudHvX64.dsc-561-
OvmfPkg/CloudHv/CloudHvX64.dsc-562-  gEfiShellPkgTokenSpaceGuid.PcdShellFileOperationSize|0x20000
--
OvmfPkg/OvmfPkgIa32.dsc-571-  # Network Pcds
OvmfPkg/OvmfPkgIa32.dsc-572-  #
OvmfPkg/OvmfPkgIa32.dsc:573:!include NetworkPkg/NetworkPcds.dsc.inc
OvmfPkg/OvmfPkgIa32.dsc-574-
OvmfPkg/OvmfPkgIa32.dsc-575-  gEfiShellPkgTokenSpaceGuid.PcdShellFileOperationSize|0x20000
--
OvmfPkg/OvmfPkgX64.dsc-603-  # Network Pcds
OvmfPkg/OvmfPkgX64.dsc-604-  #
OvmfPkg/OvmfPkgX64.dsc:605:!include NetworkPkg/NetworkPcds.dsc.inc
OvmfPkg/OvmfPkgX64.dsc-606-
OvmfPkg/OvmfPkgX64.dsc-607-  gEfiShellPkgTokenSpaceGuid.PcdShellFileOperationSize|0x20000
--
OvmfPkg/Microvm/MicrovmX64.dsc-561-  # Network Pcds
OvmfPkg/Microvm/MicrovmX64.dsc-562-  #
OvmfPkg/Microvm/MicrovmX64.dsc:563:!include NetworkPkg/NetworkPcds.dsc.inc
OvmfPkg/Microvm/MicrovmX64.dsc-564-
OvmfPkg/Microvm/MicrovmX64.dsc-565-  gEfiShellPkgTokenSpaceGuid.PcdShellFileOperationSize|0x20000
--
OvmfPkg/LoongArchVirt/LoongArchVirtQemu.dsc-374-  # Network Pcds
OvmfPkg/LoongArchVirt/LoongArchVirtQemu.dsc-375-  #
OvmfPkg/LoongArchVirt/LoongArchVirtQemu.dsc:376:!include NetworkPkg/NetworkPcds.dsc.inc
OvmfPkg/LoongArchVirt/LoongArchVirtQemu.dsc-377-
OvmfPkg/LoongArchVirt/LoongArchVirtQemu.dsc-378-  gEfiMdeModulePkgTokenSpaceGuid.PcdFlashNvStorageVariableSize         | 0x40000
--
OvmfPkg/OvmfXen.dsc-442-  # Network Pcds
OvmfPkg/OvmfXen.dsc-443-  #
OvmfPkg/OvmfXen.dsc:444:!include NetworkPkg/NetworkPcds.dsc.inc
OvmfPkg/OvmfXen.dsc-445-
OvmfPkg/OvmfXen.dsc-446-!ifdef $(DEBUG_ON_HYPERVISOR_CONSOLE)
```

Additionally, there are already include files that contain statements for checks, such as [Network.fdf.inc](https://github.com/tianocore/edk2/blob/master/NetworkPkg/Network.fdf.inc), [NetworkBuildOptions.dsc.inc](https://github.com/tianocore/edk2/blob/master/NetworkPkg/NetworkBuildOptions.dsc.inc), [NetworkComponents.dsc.inc](https://github.com/tianocore/edk2/blob/master/NetworkPkg/NetworkComponents.dsc.inc), and [NetworkDefines.dsc.inc](https://github.com/tianocore/edk2/blob/master/NetworkPkg/NetworkDefines.dsc.inc).

Comment (SaloniKasbekar) [reply NetworkPkg/NetworkPcds.dsc.inc:14]: I think thats fair. I see similar conditions for NETWORK_ENABLE used in the files you mentioned above.

**Review by SaloniKasbekar — APPROVED** [review]

**Comment by kraxel** [comment]
Creating a `NetworkPcds.dsc.inc` makes sense to me.  I'd also suggest to add checks for `NETWORK_IP4_ENABLE` and `NETWORK_IP6_ENABLE`, following existing practice of the existing `NetworkComponents.dsc.inc` file.

**Comment by Joursoir** [comment]
@kraxel please take a look, I tried to move PCDs to a separate include file

**Comment by kraxel** [comment]
Oh, I didn't realize there already is a `NetworkPkg/NetworkPcds.dsc.inc` file.
Any specific reason why you have created a new file instead of adding the PCDs to the existing file?

**Comment by Joursoir** [comment]
@kraxel Yeah, there's a specific reason why I created a new file rather than added the PCDs to the existing `NetworkPkg/NetworkPcds.dsc.inc` file.

At the moment, the NetworkPcds.dsc.inc file contains `gEfiNetworkPkgTokenSpaceGuid.PcdAllowHttpConnections` without being associated with a specific section. However, this PCD is included in the `[PcdsFixed*]` section in the DSC files.

If I add new PCDs to this include file, I would need to make sure they are in the correct sections. This would mean changing the file to include different sections like:

```
[PcdsFixedAtBuild]
gEfiNetworkPkgTokenSpaceGuid.PcdAllowHttpConnections|TRUE

[PcdsDynamicDefault]
gEfiNetworkPkgTokenSpaceGuid.PcdIPv4PXESupport|0x01
gEfiNetworkPkgTokenSpaceGuid.PcdIPv6PXESupport|0x01
```

Then, I would have to modify the DSC files that use this include file, as they reference `NetworkPcds.dsc.inc` and continue PCD declarations afterward.

Therefore, I found that separating the PCDs into `Dynamic` and `Fixed` files is a cleaner and more manageable approach

**Review by kraxel — APPROVED** [review]

**Comment by Joursoir** [comment]
A similar patch has already been merged https://github.com/tianocore/edk2/pull/6087

However, I don't believe it's the right approach to deal with network stuff just by adding _a lot_ of DEFINEs IMHO

**Review by bexcran — APPROVED** [review]

**Comment by lgao4** [comment]
@ardbiesheuvel , is there the quality risk to merge them for this table tag 202411? This patch has been reviewed before soft feature feature freeze. It can be merged for 202411.

**Comment by ardbiesheuvel** [comment]
> @ardbiesheuvel , is there the quality risk to merge them for this table tag 202411? This patch has been reviewed before soft feature feature freeze. It can be merged for 202411.

I think merging this is fine.

**Review by ardbiesheuvel — APPROVED** [review]

**Pull request merged**